### DISCUSSION TITLE
Using HyperFormula with Functional Programming and Immutable State

### DISCUSSION THREAD

**Discussion by BrianHung** [handsontable/hyperformula#1076] [Feature requests]
I'm thinking through how to integrate hyperformula with functional programming and immutable state to implement a one-directional data flow through my app.

My question is there a way to create "drafts" or "forks" of hyperformula instances where new instances re-use memory of unmodified values and only incremental create memory for modified values (structural sharing), as opposed to creating separate instances with no shared memory or values from `buildFromSheets`.

```js
const prevSheets = [[1, 2, 3]];
const prevHf = hf.buildFromSheets(prevSheets);

const nextSheets = [[1, 2, 4]]
const nextHf = hf.buildFromSheets(nextSheets);
```

Ideal API

```js
const prevSheets = [[1, 2, 3]];
const prevHf = hf.buildFromSheets(prevSheets);

// some modifications to prevSheets
const nextSheets = [[1, 2, 4]]
const nextHf = prevHf.setCellContent({pos: [0, 2], value: 4});

prevHf.getCellValue(0, 2) == 3
nextHf.getCellValue(0, 2) == 4
```

**Comment by sequba** [comment]
Hi @BrianHung!

Unfortunately, HyperFormula does not support immutability the way you described it. It is a very interesting idea, but I'm afraid it will be hard to fit it into our roadmap.

If this feature is crucial for you, feel free to work on it yourself and open a pull request to our repository. I may offer my assistance by answering the questions you might have during the development.
